forem/forem
# api v1 endpoint for creating an organization

## What type of PR is this? (check all applicable)


## Description

Following up on #19650, this PR adds a POST route under api/v1/organizations. 

Like the `destroy` action added in https://github.com/forem/forem/pull/19681, this route is restricted to site admins.

## Related Tickets & Documents



- Related Issue #19650
- Closes #19650

## QA Instructions, Screenshots, Recordings

In Uffizzi: (You will have to do this every time you want to test Uffizzi via Postman because when it goes back to sleep, it wipes out logins and api keys)

Log in as admin
Go to User Settings -> Extensions and scroll all the way down
Set a name for the user's api key, any name, doesn't matter
When the api key is generated, copy it from the reloaded page
In Postman

Set Headers -> Accept to "application/vnd.forem.api-v1+json"

Set Headers - api-key to the key you copied in Uffizzi

Make POST requests to /api/organizations to see the create errors change based on the request body.

The body tab in Postman needs params and values for name, summary, url, profile_image, slug, and tag_line, nested within a object called `organization`.

The value for profile_image can be an image url or an image file, because the controller action checks for object type and if it's a String, the value shall be cleared from `params[:organization][:profile_image]` and used to set `remote_image_url` instead.

*Authorization check:* you can look up a trusted user's login and set up an api key for them too. Using that key at this route should not work, returning a 401 error.

### UI accessibility checklist
_If your PR includes UI changes, please utilize this checklist:_

_For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?


## [optional] Are there any post deployment tasks we need to perform?
Confirming function of authorized and validated `POST api/organizations` requests ([and the other api v1 organization endpoints](https://github.com/forem/forem/pull/19681)) as expected in production.

Confirming creation and destruction are restricted to site admins as expected in production.

## [optional] What gif best describes this PR or how it makes you feel?

![colorful anthropomorphic cartoon cat (dog?) painter painting "stay creative"](https://media0.giphy.com/media/3oEduXdm2gjnrsJBOo/giphy.gif)
